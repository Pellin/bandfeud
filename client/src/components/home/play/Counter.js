import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';

import { gameOver } from '../../../actions/gameStatus';
import setCurrentPoints from '../../../actions/setCurrentPoints';

export const Counter = ({
  // band,
  difficulty,
  inGame,
  score,
  submitted,
  // onAddFailedBand,
  onGameOver,
  onSetCurrentPoints
}) => {
  const [timeLeft, setTimeLeft] = useState(0);
  useEffect(() => {
    setTimeLeft(difficulty);
  }, [difficulty]);
  useEffect(() => {
    const timer = setInterval(() => {
      if (timeLeft > 0 && inGame && !submitted) {
        setTimeLeft(timeLeft - 1);
      } else {
        // onAddFailedBand(band);
        onGameOver("Time's up, snailfinger.", score);
        clearInterval(timer);
      }
    }, 1000);
    return () => {
      clearInterval(timer);
      onSetCurrentPoints(timeLeft);
    };
  });
  return <div className="counter">{timeLeft > 0 && timeLeft}</div>;
};

const mapDispatchToProps = dispatch => ({
  // onAddFailedBand: band =>
  //   dispatch({
  //     type: 'ADD_FAILED_BAND',
  //     payload: { name: band, mode: 'Out of time' }
  //   }),
  onGameOver: (message, score) => dispatch(gameOver(message, score)),
  onSetCurrentPoints: timeLeft => dispatch(setCurrentPoints(timeLeft))
});

export default connect(undefined, mapDispatchToProps)(Counter);
