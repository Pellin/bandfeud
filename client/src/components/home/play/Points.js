import React, { useState, useEffect } from 'react';

const Points = ({ points }) => {
  const [show, setShow] = useState(true);
  useEffect(() => {
    const timeout = setTimeout(() => {
      setShow(false);
    }, 400);
    return () => {
      clearTimeout(timeout);
    };
  });
  return (
    <>
      {points && (
        <div className={show ? 'points' : 'points-hide'}>+{points}</div>
      )}
    </>
  );
};

export default Points;
