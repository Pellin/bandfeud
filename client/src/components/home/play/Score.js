import React, { useState, useEffect } from 'react';

const Score = props => {
  const [displayedScore, setDisplayedScore] = useState(0);
  useEffect(() => {
    if (displayedScore !== props.score) {
      const ticker = setInterval(() => {
        setDisplayedScore(displayedScore + 1);
        return clearInterval(ticker);
      }, 10);
    }
  }, [props.score, displayedScore]);
  return <>{displayedScore}p</>;
};

export default Score;
