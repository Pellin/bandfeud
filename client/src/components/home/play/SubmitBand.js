import React, { useState } from 'react';
import { connect } from 'react-redux';
import { motion } from 'framer-motion';

import Counter from './Counter';

import checkBand from '../../../actions/checkBand';
import { setDifficulty } from '../../../actions/setDifficulty';

export const SubmitBand = props => {
  const [band, setBand] = useState('');
  const [iosPlaceHold, setIosPlaceHold] = useState('press to type');

  const onRemovePlaceholder = () => {
    setIosPlaceHold('');
  };
  const onBandChange = e => {
    let bandName = e.target.value;
    if (
      bandName.match(/^[a-zåäö0-9/\-&-,.\s]*$/) &&
      !bandName.match(/^[.,-\s]/)
    ) {
      setBand(bandName);
    }
  };
  const onSubmit = e => {
    e.preventDefault();
    if (band.length < 2 || !band.match(/[\wåäö]$/)) return;

    props.onCheckBand(
      band,
      props.previous,
      props.used,
      props.score,
      props.difficulty
    );
    setBand('');
    props.onSetDifficulty(props.used.length);
  };
  return (
    <>
      {props.os !== 'desktop' && (
          <Counter
            addFailedBand={props.onAddFailedBand}
            submitted={props.submitted}
            inGame={props.inGame}
            used={props.used}
            difficulty={props.difficulty}
            score={props.score}
          />
      )}
      <form
        className={
          props.os === 'iOS' ? 'band-input-form-ios' : 'band-input-form'
        }
        onSubmit={onSubmit}
      >
        <input
          className={
            band.length > 25 && props.os !== 'desktop'
              ? 'band-input-long'
              : 'band-input'
          }
          autoCapitalize="none"
          autoComplete="off"
          autoCorrect="off"
          autoFocus={props.os !== 'desktop' ? false : true}
          maxLength="50"
          onChange={onBandChange}
          onBlur={props.os !== 'desktop' && band.length > 1 ? onSubmit : null}
          onFocus={() => onRemovePlaceholder()}
          placeholder={props.os !== 'desktop' ? iosPlaceHold : undefined}
          spellCheck={false}
          type="text"
          value={band}
        />
      </form>
      {props.previous.previous2 ? (
        <div className="previous">
          {props.previous.previous1.toUpperCase() +
            ' or ' +
            props.previous.previous2.toUpperCase()}
        </div>
      ) : (
        <motion.div
          className="previous"
          animate={{ scale: [null, 1.3, 1] }}
          transition={{ duration: 1 }}
        >
          {props.previous.previous1.toUpperCase()}
        </motion.div>
      )}
    </>
  );
};

const mapDispatchToProps = dispatch => ({
   onAddFailedBand: band =>
    dispatch({
      type: 'ADD_FAILED_BAND',
      payload: { name: band, mode: 'Out of time' }
    }),
  onCheckBand: (band, previous, used, score, difficulty) =>
    dispatch(checkBand(band, previous, used, score, difficulty)),
  onSetDifficulty: usedLength => dispatch(setDifficulty(usedLength))
});

export default connect(
  undefined,
  mapDispatchToProps
)(SubmitBand);
