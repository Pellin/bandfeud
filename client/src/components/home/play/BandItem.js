import React from 'react';
import Points from '../play/Points';

const BandItem = ({ name, url, points }) => {
  const fireReg = new RegExp('Firefox');
  const isFirefox = fireReg.test(navigator.userAgent);
  return (
    <div className="band-item">
      <div className="image-container">
        <div className="band-item-image">
          <img className="img" src={url} alt={'"' + { name } + '"'} />
        </div>
        {points && <Points points={points} />}
      </div>
      <div className={isFirefox ? 'title-flow-firefox' : 'title-flow'}>
        <div className="band-item-title-container">
          <div className="band-item-title">{name}</div>
        </div>
        
      </div>
    </div>
  );
};

export default BandItem;
