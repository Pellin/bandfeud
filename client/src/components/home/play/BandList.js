import React, { Component } from 'react';

import BandItem from './BandItem';

class BandList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      left: '0px',
      children: 0
    };
  }
  componentDidMount = () => {
    window.addEventListener('resize', this.setCenter);
    this.setState(() => ({ children: this.props.bands.length }));
    this.setCenter();
  };
  componentDidUpdate = () => {
    if (this.props.bands.length !== this.state.children) {
      this.setState(() => ({
        children: this.props.bands.length
      }));
      this.setState(() => ({
        left: parseInt(this.state.left.slice(0, -2)) - 210 + 'px'
      }));
    }
  };
  componentWillUnmount = () => {
    window.removeEventListener('resize', this.setCenter);
  };
  setCenter = () => {
    this.setState(() => ({
      left: window.innerWidth / 2 - this.props.bands.length * 210 + 105 + 'px'
    }));
  };
  render() {
    return (
      <div
        style={{
          alignItems: 'center',
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'center',
          top: '5rem',
          marginBottom: '1rem',
          left: this.state.left,
          position: 'absolute',
          transition: 'left, 1s'
        }}
      >
        {this.props.bands.map(band => (
          <BandItem key={band.name} name={band.name} url={band.url} points={band.points}/>
        ))}
      </div>
    );
  }
}

export default BandList;
