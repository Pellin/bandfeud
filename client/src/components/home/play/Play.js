import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { motion } from 'framer-motion';

import SubmitBand from './SubmitBand';
import BandList from './BandList';
import MessageBox from '../MessageBox';
import Score from './Score';
import BfGoA from '../../../icons/BfGoA';

import SubmitHighscore from '../../highscores/SubmitHighscore';
import { gameAborted } from '../../../actions/gameStatus';

export const Play = props => {
  const [isFirefox, setIsFirefox] = useState(false);
  const abortGame = () => {
    if (props.inGame) {
      props.onGameAborted();
    }
  };
  useEffect(() => {
    const listener =
      props.os === 'desktop' &&
      (/Firefox/gi.test(navigator.userAgent) ||
        /Chrome/gi.test(navigator.userAgent))
        ? 'visibilitychange'
        : 'blur';
    window.addEventListener(listener, abortGame);
    return () => {
      window.removeEventListener(listener, abortGame);
      props.releaseButton();
    };
  });
  useEffect(() => {
    const fireReg = new RegExp('Firefox');
    setIsFirefox(fireReg.test(navigator.userAgent));
  }, []);
  return (
    <>
      {props.showGameOver ? (
        <>
          <div className="game-over-container">
            <motion.div
              className="game-over-logo"
              initial={{ scale: 3, y: 'calc(75vh)', opacity: 0 }}
              animate={{ scale: 1, rotate: [30, 0], y: 0, opacity: 1 }}
              transition={{ ease: 'backOut', duration: 1, type: 'tween' }}
            >
              <BfGoA />
            </motion.div>
          </div>
          <div className="footer">
            <MessageBox message={props.message} />
          </div>
        </>
      ) : (
        <div className="view-holder">
          {props.isHighscore ? (
            <SubmitHighscore
              bands={props.bands}
              difficulty={props.difficulty}
              history={props.history}
              inGame={props.inGame}
              score={props.score}
              submitted={props.submitted}
              used={props.used}
            />
          ) : (
            <div
              className={
                isFirefox ? 'game-container-firefox' : 'game-container'
              }
            >
              <div className="band-list">
                <BandList bands={props.bands} />
              </div>
              {props.inGame && props.os !== 'desktop' && (
                <div className="score">
                  <Score score={props.score} />
                </div>
              )}
              <div className="footer">
                {props.message ? (
                  <MessageBox message={props.message} />
                ) : (
                  <div className="submit-band">
                    {props.inGame && !props.submitted && !props.message && (
                      <SubmitBand
                        bandBank={props.bandBank}
                        difficulty={props.difficulty}
                        inGame={props.inGame}
                        previous={props.previous}
                        score={props.score}
                        submitted={props.submitted}
                        used={props.used}
                        os={props.os}
                      />
                    )}
                  </div>
                )}
              </div>
            </div>
          )}
        </div>
      )}
    </>
  );
};

const mapDispatchToProps = dispatch => ({
  onGameAborted: () => dispatch(gameAborted())
});

const mapStateToProps = state => ({
  bandBank: state.bandBank,
  used: state.used,
  submitted: state.submitted,
  previous: state.previous,
  inGame: state.inGame,
  isHighscore: state.isHighscore,
  difficulty: state.difficulty,
  score: state.score,
  showGameOver: state.showGameOver,
  bands: state.bands,
  message: state.message,
  os: state.os
});

export default connect(mapStateToProps, mapDispatchToProps)(Play);
