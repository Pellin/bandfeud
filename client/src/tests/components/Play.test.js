import React from 'react';
import { shallow } from 'enzyme';

import { Play } from '../../components/home/play/Play';

let used, previous;

beforeEach(() => {
  used = ['one', 'two', 'three'];
  previous = 'a';
});

it('should render Play correctly with get ready message', () => {
  const wrapper = shallow(
    <Play
      used={[]}
      submitted={true}
      previous={previous}
      inGame={true}
      isHighscore={false}
      difficulty={20}
      score={0}
      bands={[]}
      message={'Get ready...'}
    />
  );

  expect(wrapper).toMatchSnapshot();
});

it('should show submitform when submitted is falsy', () => {
  const wrapper = shallow(
    <Play
      used={[]}
      submitted={false}
      previous={previous}
      inGame={true}
      isHighscore={false}
      difficulty={20}
      score={0}
      bands={[]}
      message={''}
    />
  );

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('.submit-band').children().length).toBe(1);
});

it('should not show submitform and counter when submitted is true', () => {
  const wrapper = shallow(
    <Play
      used={[]}
      submitted={true}
      previous={previous}
      inGame={true}
      isHighscore={false}
      difficulty={20}
      score={0}
      bands={[]}
      message={''}
    />
  );

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('.submit-band').children().length).toBe(0);
  expect(wrapper.find('.counter-container').children().length).toBe(0);
});

it('should show bands in band-list', () => {
  const wrapper = shallow(
    <Play
      used={used}
      submitted={false}
      previous={'r'}
      inGame={true}
      isHighscore={false}
      difficulty={20}
      score={76}
      bands={[
        { name: 'one', url: 'www.1img.com' },
        { name: 'two', url: 'www.2img.com' },
        { name: 'three', url: 'www.3img.com' },
        { name: 'four', url: 'www.4img.com' }
      ]}
      message={''}
    />
  );

  expect(wrapper).toMatchSnapshot();
  expect(
    wrapper
      .find('.band-list')
      .children()
      .props('bands')
      .bands
  ).toEqual([
    { name: 'one', url: 'www.1img.com' },
    { name: 'two', url: 'www.2img.com' },
    { name: 'three', url: 'www.3img.com' },
    { name: 'four', url: 'www.4img.com' }
  ]);
});

it('should show SubmitHighscore instead if state is right', () => {
  const wrapper = shallow(
    <Play
      history={() => {}}
      used={used}
      submitted={false}
      previous={'r'}
      inGame={true}
      isHighscore={true}
      difficulty={20}
      score={76}
      bands={[
        { name: 'one', url: 'www.1img.com' },
        { name: 'two', url: 'www.2img.com' },
        { name: 'three', url: 'www.3img.com' },
        { name: 'four', url: 'www.4img.com' }
      ]}
      message={''}
    />
  );

  expect(wrapper).toMatchSnapshot();
});
