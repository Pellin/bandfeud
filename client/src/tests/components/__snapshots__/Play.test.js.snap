// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not show submitform and counter when submitted is true 1`] = `
<Fragment>
  <div
    className="view-holder"
  >
    <div
      className="game-container"
    >
      <div
        className="band-list"
      >
        <BandList
          bands={Array []}
        />
      </div>
      <div
        className="score"
      >
        <Score
          score={0}
        />
      </div>
      <div
        className="footer"
      >
        <div
          className="submit-band"
        />
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`should render Play correctly with get ready message 1`] = `
<Fragment>
  <div
    className="view-holder"
  >
    <div
      className="game-container"
    >
      <div
        className="band-list"
      >
        <BandList
          bands={Array []}
        />
      </div>
      <div
        className="score"
      >
        <Score
          score={0}
        />
      </div>
      <div
        className="footer"
      >
        <MessageBox
          message="Get ready..."
        />
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`should show SubmitHighscore instead if state is right 1`] = `
<Fragment>
  <div
    className="view-holder"
  >
    <Connect(SubmitHighscore)
      bands={
        Array [
          Object {
            "name": "one",
            "url": "www.1img.com",
          },
          Object {
            "name": "two",
            "url": "www.2img.com",
          },
          Object {
            "name": "three",
            "url": "www.3img.com",
          },
          Object {
            "name": "four",
            "url": "www.4img.com",
          },
        ]
      }
      difficulty={20}
      history={[Function]}
      inGame={true}
      score={76}
      submitted={false}
      used={
        Array [
          "one",
          "two",
          "three",
        ]
      }
    />
  </div>
</Fragment>
`;

exports[`should show bands in band-list 1`] = `
<Fragment>
  <div
    className="view-holder"
  >
    <div
      className="game-container"
    >
      <div
        className="band-list"
      >
        <BandList
          bands={
            Array [
              Object {
                "name": "one",
                "url": "www.1img.com",
              },
              Object {
                "name": "two",
                "url": "www.2img.com",
              },
              Object {
                "name": "three",
                "url": "www.3img.com",
              },
              Object {
                "name": "four",
                "url": "www.4img.com",
              },
            ]
          }
        />
      </div>
      <div
        className="score"
      >
        <Score
          score={76}
        />
      </div>
      <div
        className="footer"
      >
        <div
          className="submit-band"
        >
          <Connect(SubmitBand)
            bandBank={
              Array [
                Object {
                  "name": "",
                  "url": "",
                },
                Object {
                  "name": "",
                  "url": "",
                },
                Object {
                  "name": "",
                  "url": "",
                },
              ]
            }
            difficulty={20}
            inGame={true}
            previous="r"
            score={76}
            submitted={false}
            used={
              Array [
                "one",
                "two",
                "three",
              ]
            }
          />
        </div>
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`should show submitform when submitted is falsy 1`] = `
<Fragment>
  <div
    className="view-holder"
  >
    <div
      className="game-container"
    >
      <div
        className="band-list"
      >
        <BandList
          bands={Array []}
        />
      </div>
      <div
        className="score"
      >
        <Score
          score={0}
        />
      </div>
      <div
        className="footer"
      >
        <div
          className="submit-band"
        >
          <Connect(SubmitBand)
            bandBank={
              Array [
                Object {
                  "name": "",
                  "url": "",
                },
                Object {
                  "name": "",
                  "url": "",
                },
                Object {
                  "name": "",
                  "url": "",
                },
              ]
            }
            difficulty={20}
            inGame={true}
            previous="a"
            score={0}
            submitted={false}
            used={Array []}
          />
        </div>
      </div>
    </div>
  </div>
</Fragment>
`;
